<a href="https://colab.research.google.com/github/coding-dojo-data-science/machine-learning-practice-solutions/blob/main/Classification_Tree_Exercise_(Practice)_Solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Classification Tree Exercise (Practice) - Solution

- [Assignment Link](https://login.codingdojo.com/m/213/7198/63577)
- Solution by: James Irving
- 11/23/21

# Instructions


This task uses a breast cancer data set that can be found [here.](https://drive.google.com/file/d/1Zms1RfgkWrTp7S6_BFpeELYyvX1s2FSN/view?usp=sharing)  The target vector is the diagnosis as either malignant (M) or benign (B).  

Your task is to use a 

1) decision tree classifier 

2) bagging classifier and a 

3) random forest classifier 

to obtain the highest accuracy possible on the test set. 

You may want to refer to the previous chapter to review bagging and random forest as they were explained with regression. They work very similar in a classification problem, but the final prediction is a majority vote of which class instead of an average of continuous values.  

Note, the following code can be used to import the classification tree models required:

```python
from sklearn.tree import DecisonTreeClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import RandomForest Classifier
```

# Solution

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, BaggingClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score

In [3]:
## Load in the dataset from url
fpath = "https://docs.google.com/spreadsheets/d/e/2PACX-1vR3X81gX_Zlbs4j244cQbBe04CTZbJ-Dl6EwNzKZRFngs9XZu_9dzVDuzzpZJgCj6xqS8oHR8K3eqr9/pub?output=csv"

df = pd.read_csv(fpath, index_col=0)
df

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
...            ...          ...           ...             ...        ...   
926424           M        21.56         22.39          142.00     1479.0   
926682           M        20.13         28.25          131.20     1261.0   
926954           M        16.60         28.08          108.30      858.1   
927241           M        20.60         29.33          140.10     1265.0   
92751            B         7.76         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
...                   ...               ...             ...   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  radius_worst  \
id                                            ...                 
842302                0.14710         0.2419  ...        25.380   
842517                0.07017         0.1812  ...        24.990   
84300903              0.12790         0.2069  ...        23.570   
84348301              0.10520         0.2597  ...        14.910   
84358402              0.10430         0.1809  ...        22.540   
...                       ...            ...  ...           ...   
926424                0.13890         0.1726  ...        25.450   
926682                0.09791         0.1752  ...        23.690   
926954                0.05302         0.1590  ...        18.980   
927241                0.15200         0.2397  ...        25.740   
92751                 0.00000         0.1587  ...         9.456   

          texture_worst  perimeter_worst  area_worst  smoothness_worst  \
id                                                                       
842302            17.33           184.60      2019.0           0.16220   
842517            23.41           158.80      1956.0           0.12380   
84300903          25.53           152.50      1709.0           0.14440   
84348301          26.50            98.87       567.7           0.20980   
84358402          16.67           152.20      1575.0           0.13740   
...                 ...              ...         ...               ...   
926424            26.40           166.10      2027.0           0.14100   
926682            38.25           155.00      1731.0           0.11660   
926954            34.12           126.70      1124.0           0.11390   
927241            39.42           184.60      1821.0           0.16500   
92751             30.37            59.16       268.6           0.08996   

          compactness_worst  concavity_worst  concave points_worst  \
id                                                                   
842302              0.66560           0.7119                0.2654   
842517              0.18660           0.2416                0.1860   
84300903            0.42450           0.4504                0.2430   
84348301       

In [4]:
## Make X and y 
target = 'diagnosis'
X = df.drop(columns=target).copy()
y = df[target].copy()


In [5]:
## train-test-split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
X_train

radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                
8913            12.890         13.12           81.89      515.9   
915691          13.400         20.52           88.64      556.7   
904689          12.960         18.29           84.18      525.2   
9110732         17.750         28.03          117.30      981.6   
881046502       20.580         22.14          134.70     1290.0   
...                ...           ...             ...        ...   
859711           8.888         14.64           58.79      244.0   
863031          11.640         18.33           75.17      412.5   
8910721         14.290         16.82           90.30      632.6   
908489          13.980         19.62           91.12      599.5   
862965          12.180         20.52           77.22      458.7   

           smoothness_mean  compactness_mean  concavity_mean  \
id                                                             
8913               0.06955           0.03729         0.02260   
915691             0.11060           0.14690         0.14450   
904689             0.07351           0.07899         0.04057   
9110732            0.09997           0.13140         0.16980   
881046502          0.09090           0.13480         0.16400   
...                    ...               ...             ...   
859711             0.09783           0.15310         0.08606   
863031             0.11420           0.10170         0.07070   
8910721            0.06429           0.02675         0.00725   
908489             0.10600           0.11330         0.11260   
862965             0.08013           0.04038         0.02383   

           concave points_mean  symmetry_mean  fractal_dimension_mean  ...  \
id                                                                     ...   
8913                   0.01171         0.1337                 0.05581  ...   
915691                 0.08172         0.2116                 0.07325  ...   
904689                 0.01883         0.1874                 0.05899  ...   
9110732                0.08293         0.1713                 0.05916  ...   
881046502              0.09561         0.1765                 0.05024  ...   
...                        ...            ...                     ...  ...   
859711                 0.02872         0.1902                 0.08980  ...   
863031                 0.03485         0.1801                 0.06520  ...   
8910721                0.00625         0.1508                 0.05376  ...   
908489                 0.06463         0.1669                 0.06544  ...   
862965                 0.01770         0.1739                 0.05677  ...   

           radius_worst  texture_worst  perimeter_worst  area_worst  \
id                                                                    
8913             13.620          15.54            87.40       577.0   
915691           16.410          29.66           113.30       844.4   
904689           14.130          24.61            96.31       621.9   
9110732          21.530          38.54           145.40      1437.0   
881046502        23.240          27.84           158.30      1656.0   
...                 ...            ...              ...         ...   
859711            9.733          15.67            62.56       284.4   
863031           13.140          29.26            85.51       521.7   
8910721          14.910          20.65            94.44       684.6   
908489           17.040          30.80           113.90       869.3   
862965           13.340          32.84            84.58       547.8   

           smoothness_worst  compactness_worst  concavity_worst  \
id                                                                
8913                0.09616            0.11470          0.11860   
915691              0.15740            0.38560          0.51060   
904689              0.09329            0.23180          0.16040   
9110732             0.14010     

### Modeling

In [6]:
## Evaluation function
def evluate_classification(model,X_train,X_test,y_train,y_test):
    """Uses model.score to display the accuracy scores for training and test splits"""
    print(f"Training Accuracy: {model.score(X_train, y_train):.3f}")
    print(f"Test Accuracy: {model.score(X_test, y_test):.3f}")


In [7]:
## Make, fit, and evalaute Decision Tree
tree_clf = DecisionTreeClassifier(random_state=42)
tree_clf.fit(X_train, y_train)

evluate_classification(tree_clf, X_train, X_test, y_train, y_test)

Training Accuracy: 1.000
Test Accuracy: 0.951


In [8]:
## Make, fit, and evalaute BaggingClassifier
bag_clf = BaggingClassifier(random_state=42)
bag_clf.fit(X_train, y_train)

evluate_classification(bag_clf, X_train, X_test, y_train, y_test)

Training Accuracy: 0.993
Test Accuracy: 0.951


In [9]:
## Make, fit, and evalaute BaggingClassifier
rf_clf = RandomForestClassifier(random_state=42)
rf_clf.fit(X_train, y_train)

evluate_classification(rf_clf, X_train, X_test, y_train, y_test)

Training Accuracy: 1.000
Test Accuracy: 0.965


> The best model, according to the test accuracy, is the RandomForestClassifier: 
- Training accuracy: 1.0
- Test accuracy: 0.965